### ISSUE TITLE
RcisTarget

### ISSUE THREAD

**Issue by s-aibar** [Bioconductor/Contributions#502]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/aertslab/RcisTarget.hg19.motifDBs.cisbpOnly.500bp

Confirm the following by editing each check box to '[x]'

- [x] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [x] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [x] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [x] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [x] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [x] The 'devel' branch for new packages and features.
- [x] The stable 'release' branch, made available every six
      months, for bug fixes.
- [x] _Bioconductor_ version control using [Git][5]
  (optionally [via GitHub][6]).

For help with submitting your package, please subscribe and post questions
to the [bioc-devel][4] mailing list.

[1]: https://bioconductor.org/developers/package-guidelines/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/git/
[6]: http://bioconductor.org/developers/how-to/git/sync-existing-repositories/

**Comment by s-aibar** [comment]
This are the "databases" to use with the main package, RcisTarget (https://github.com/aertslab/RcisTarget).

Following the guidelines, I will add the "AdditionalPackage" line once it has passed the build without errors and warnings, and "completed the 'moderation' step".

**Comment by s-aibar** [comment]
AdditionalPackage: https://github.com/aertslab/RcisTarget

**Comment by s-aibar** [comment]
Hello @LiNk-NY ,

The current warning in RcisTarget seems to come from package GSEABase (it only occurs in devel):
https://bioconductor.org/checkResults/devel/bioc-LATEST/GSEABase/malbec1-checksrc.html

I have tried to import only the class/methods that RcisTarget uses from GSEABase, but it still triggers the warning. Any tips on how to proceed? (Should I do something, or just wait for GSEABase to solve it?)

Thank you,
S. Aibar

**Comment by s-aibar** [comment]
(I have temporarily removed the dependency to GSEABase. I will add it back once the warning is solved, or you tell me otherwise...)

**Comment by LiNk-NY** [comment]
Hi Sara, @s-aibar
Apologies for the delay in getting back to you. 
I will be looking at your package later today.
Best regards,
Marcel

**Comment by LiNk-NY** [comment]
Hi Sara, @s-aibar 

Unfortunately, the review for your package had to be pushed back. 
I had to work on some fixes for a bunch of packages. 
Thank you for your patience. 

Please see the review below. 

If you have any questions, feel free to contact me.

Best regards,
Marcel

----

# RcisTarget #502

## DESCRIPTION

* Please add a `Depends` field `R >= (3.4.1)`
* Include a `BugReports` URL pointing to the issues
location on GitHub
* Include `LazyLoad: yes` field and value 

## NAMESPACE

* Use `roxygen2` directives to import functions from
`data.table`
* The above can be done by adding `@import` and a `NULL`
after the roxygen block
* Avoid manual generation of the `NAMESPACE` 

## R

Major:
* `RcisTarget` your main function, should use existing
infrastructure rather than make use of `data.table`
* Perhaps you can pick an existing class and extend it?
* Consider using `BiocParallel` for functions that require
parallelization
* Please keep the support of `GSEABase` enabled despite the
warnings those should clear up. If not, I can contact the
person working on that package

Grammar:
* Avoid using `sapply` as it is not robust to input changes.
* Use `vapply` instead where possible.
* Avoid growing a vector, use the allocate and fill method. 
* Example: `in000 <- NULL` then `in000 <- sapply(...)` in this
case you can just use the `lapply`/`vapply` functions and not
worry about allocating memory which is done internally.
* It might be better to create a method with signature "ANY"
and direct that to `.getSignificantGenes` so that you can
just call `callNextMethod` for all methods after extracting
the desired `geneSet` from that particular class
* Use `requireNamespace` to test for package availability
rather than `installed.packages`
* It seems like `fakeDatabase` should be an internal function
and not exported, or saved as a script
* Do not call empty `stop` functions, provide a useful error
message
* Avoid `0:length(x)` use the more robust `seq_along` or
`seq_len`

Minor: 
* Logical to Numeric index switching not necessary most of
the time (e.g., `which(nes >= nesThreshold)` can be
simplified to `nes >= nesThreshold`. 

## vignettes

* Keep vignette lines under 80 column widths
* Looks good

**Comment by s-aibar** [comment]
Hello @LiNk-NY ,

Thank you for the review, I guess it is a quite busy moment right before the new release... 

I am already working on your comments, and I will upload a revised version tomorrow. However,  I have some doubts regarding some of the points:

> *RcisTarget your main function, should use existing infrastructure rather than make use of data.table. Perhaps you can pick an existing class and extend it?*

The data.table is mostly used to store the motif rankings (which are usually a ‘table’ of 20k genes by 20k motifs, and need to be processed in every analysis). In the first trials for the package I just used a matrix or data.frame (to include in a summarizedExperiment), but it turned out to be quite slow, and using data.table instead was faster. However, I think summarizeExperiment does not accept a data.table as assay… does it? If you have any suggestions on how to improve this implementation (i.e. any specific Bioc class?),  they are welcome… 

> It might be better to create a method with signature "ANY" and direct that to .getSignificantGenes so that you can just call callNextMethod for all methods after extracting
the desired geneSet from that particular class

The current implementation of the methods/signatures is based on the suggestions for a previous package (https://github.com/Bioconductor/Contributions/issues/321#issuecomment-307092993). I don’t mind changing it to use `callNextMethod`, but I don’t clearly see how to do it. Would you mind pointing me towards some examples/guide?

> *Avoid growing a vector, use the allocate and fill method.
Example: in000 <- NULL then in000 <- sapply(...) in this case you can just use the lapply/vapply functions and not worry about allocating memory which is done internally.*

I’m not sure which parts of the code “grow a vector”. Do you mean when it adds some extra fields, for example to a return object? E.g.:
```
  ret <- list(enrStats=enrStats) 
  if("geneList" %in% genesFormat)
    ret <- c(ret, enrichedGenes=list(enrichedGenes))
```

> *It seems like fakeDatabase should be an internal function and not exported, or saved as a script*

This function is certainly not a “core function”, but rather to run the examples quickly. 
However, if it is not exported, the users will not be able to run the examples. Will they? 
And if it is saved as a script it will require adding one/two lines of code in each example to load it… Is it worth it?

> *Please add a Depends field R >= (3.4.1)*

Is it really needed that it is 3.4.1? I usually use the package on a server with R 3.2 (I cannot upgrade R/Bioc myself) and it works fine… If it requires having a higher R version, it might unnecessarily restrict the use to other people in similar situations…?

---
*Already done (but I will do some more tests before pushing it to Github):*

## DESCRIPTION
- [x] Include a BugReports URL pointing to the issues
location on GitHub

## NAMESPACE
- [x] Use roxygen2 directives to import functions from
data.table
The above can be done by adding @import and a NULL
after the roxygen block
- [x] Avoid manual generation of the NAMESPACE
## R
Major:
- [ ] Consider using BiocParallel for functions that require
parallelization
**I will**
- [x] Please keep the support of GSEABase enabled despite the
warnings those should clear up. If not, I can contact the
person working on that package

Grammar:
- [x] Avoid using sapply as it is not robust to input changes. Use vapply instead where possible.
- [x] Use requireNamespace to test for package availability
rather than installed.packages
- [x] Do not call empty stop functions, provide a useful error
message
- [x] Avoid 0:length(x) use the more robust seq_along orseq_len

Minor:
- [x] Logical to Numeric index switching not necessary most of
the time (e.g., which(nes >= nesThreshold) can be
simplified to nes >= nesThreshold.

## vignettes
- [x] Keep vignette lines under 80 column widths
**All code lines are now under 80 ch. The text explanations are kept as paragraphs (the markdown/html editor/viewer will wrap them…) Is that ok?**

**Comment by vjcitn** [comment]
Consider HDF5Array for your big data.  Compare to data.table and let us
know how it goes.

**Comment by PeteHaitch** [comment]
> The data.table is mostly used to store the motif rankings (which are usually a ‘table’ of 20k genes by 20k motifs, and need to be processed in every analysis). In the first trials for the package I just used a matrix or data.frame (to include in a summarizedExperiment), but it turned out to be quite slow, and using data.table instead was faster. However, I think summarizeExperiment does not accept a data.table as assay… does it? If you have any suggestions on how to improve this implementation (i.e. any specific Bioc class?), they are welcome…

What sort of operations do you need to perform on the data? **data.table** is certainly a wonderful package with blazing fast performance. However, as your data are small enough to store in-memory and **if** the operations can be written in a way to use (possibly grouped) row/column summaries (e.g., `rowSums()`, `rowMeans()`), then I would be surprised if comparable performance can't be achieved by storing your data in a _matrix_ and using the [**matrixStats**](https://cran.r-project.org/package=matrixStats) package (or even base-R functionality) combined with `tapply()`/`lapply()` to do any grouped operations. I'm happy to help, perhaps with a proof of concept, if you can describe some of the sort of operations that were quite slow in your initial implementation.

**Comment by s-aibar** [comment]
AdditionalPackage: https://github.com/aertslab/RcisTarget.hg19.motifDBs.cisbpOnly.500bp

**Comment by s-aibar** [comment]
Hello @LiNk-NY ,

I have finally updated the rankings class to use SummarizedExperiment (based on a matrix) instead of a data.table. I think the main changes are now ready.

These are the comments that are still left (most of them from the 'grammar' section): 

> It might be better to create a method with signature "ANY" and direct that to .getSignificantGenes so that you can just call callNextMethod for all methods after extracting
the desired geneSet from that particular class

The current implementation of the methods/signatures is based on the suggestions for a previous package (https://github.com/Bioconductor/Contributions/issues/321#issuecomment-307092993). I don’t mind changing it to use `callNextMethod`, but I don’t clearly see how to do it. Would you mind pointing me towards some examples/guide?

> *Avoid growing a vector, use the allocate and fill method.
Example: in000 <- NULL then in000 <- sapply(...) in this case you can just use the lapply/vapply functions and not worry about allocating memory which is done internally.*

I’m not sure which parts of the code “grow a vector”. Do you mean when it adds some extra fields, for example to a return object? E.g.:
```
  ret <- list(enrStats=enrStats) 
  if("geneList" %in% genesFormat)
    ret <- c(ret, enrichedGenes=list(enrichedGenes))
```

> *It seems like fakeDatabase should be an internal function and not exported, or saved as a script*

This function is certainly not a “core function”, but rather to run the examples quickly. 
However, if it is not exported, the users will not be able to run the examples. Will they? 
And if it is saved as a script it will require adding one/two lines of code in each example to load it… Is it worth it?

> *Please add a Depends field R >= (3.4.1)*

Is it really needed that it is 3.4.1? I usually use the package on a server with R 3.2 (I cannot upgrade R/Bioc myself) and it works fine… If it requires having a higher R version, it might unnecessarily restrict the use to other people in similar situations…?

> Consider using BiocParallel for functions that require parallelization

I will, but I'm not sure whether I will have time within the next couple of days).


---
*Checklist*

## DESCRIPTION
- [x] Include a BugReports URL pointing to the issues
location on GitHub

## NAMESPACE
- [x] Use roxygen2 directives to import functions from data.table
The above can be done by adding @import and a NULL after the roxygen block
- [x] Avoid manual generation of the NAMESPACE
## R
Major:
- [x] RcisTarget your main function, should use existing infrastructure rather than make use of data.table. Perhaps you can pick an existing class and extend it?
- [ ] Consider using BiocParallel for functions that require parallelization
- [x] Please keep the support of GSEABase enabled despite the
warnings those should clear up. If not, I can contact the
person working on that package

Grammar:
- [x] Avoid using sapply as it is not robust to input changes. Use vapply instead where possible.
- [x] Use requireNamespace to test for package availability
rather than installed.packages
- [x] Do not call empty stop functions, provide a useful error
message
- [x] Avoid 0:length(x) use the more robust seq_along orseq_len
- [ ] Avoid growing a vector, use the allocate and fill method. Example: in000 <- NULL then in000 <- sapply(...) in this case you can just use the lapply/vapply functions and not worry about allocating memory which is done internally.
- [ ] It might be better to create a method with signature "ANY" and direct that to .getSignificantGenes so that you can just call callNextMethod for all methods after extracting the desired geneSet from that particular class
- [ ] It seems like fakeDatabase should be an internal function and not exported, or saved as a script

Minor:
- [x] Logical to Numeric index switching not necessary most of the time (e.g., which(nes >= nesThreshold) can be simplified to nes >= nesThreshold.

## vignettes
- [x] Keep vignette lines under 80 column widths
**All code lines are now under 80 ch. The text explanations are kept as paragraphs (the markdown/html editor/viewer will wrap them…) Is that ok?**

**Comment by LiNk-NY** [comment]
Hi Sara, @s-aibar 
You can set an any method via:
```
setMethod("getSignificantGenes", "ANY", .getSignifcantGenes)
```
then all your other methods can just be:
```
setMethod("getSignificantGenes", "GeneSetCollection", 
function(geneSet, rankings, signifRankingNames=NULL, method="iCisTarget",
    maxRank=5000, plotCurve=FALSE, genesFormat=c("geneList", "incidMatrix"),
    nCores=1, digits=3, nMean=10) {
    print(plotCurve)
    if(length(geneSet)>1) stop("Provide only one gene set.")
    geneSet <- unlist(GSEABase::geneIds(geneSet))
    callNextMethod(geneSets=geneSet) ## etc... might require rest of arguments
})
```

The parts of the code that grow a vector have that pattern of `obj <- c(obj, newData)`.
I would recommend that you pre-allocate an object and then fill it. Small example below:

```
chars <- vector("character", 5L)
chars <- letters[1:5]
```

The `fakeDatabase` function does not need to be exported and it should not be exported.
It should be an internal function denoted by a period (`.fakeDatabase`). You should be able
to use it internally when you need to. Why not use a small subset of a real database? 
You could save this dataset in the package and use it for examples. 


With regards to the `R` version, it is highly recommended that you test building your package
with the latest R version found in the Bioconductor Build System, `3.4.2 patched`. You can 
keep the lower `3.2` version in the `DESCRIPTION` too. Another option is to update your
server's R version.

Regads,
Marcel

**Comment by LiNk-NY** [comment]
Hi Sara, @s-aibar 
Any updates on the status of the package? 
Thank you,
Marcel

**Comment by LiNk-NY** [comment]
Hi Sara, @s-aibar 
Please advise on updates to the repository.
Otherwise, I will close the issue. 
Thank you. 
Regards,
Marcel

**Comment by s-aibar** [comment]
Hello,
Sorry for the delay, I already did most of the updates to the packabe, but I'm waiting for the R version of the servers to be updated to test it with the new version. Hopefully it will not take much longer. 
Regards,
Sara

**Comment by LiNk-NY** [comment]
Hi Sara, @s-aibar 

The `R.version` on the SPB has been updated to `3.5`. You can bump the version and test the package. 

Best regards,
Marcel

**Comment by LiNk-NY** [comment]
Hi Sara, @s-aibar 
Any updates on bumping the version of the package? 
Thanks,
Marcel

**Comment by LiNk-NY** [comment]
Hi Sara, @s-aibar 

Please respond. 
Otherwise, I will close the issue due to inactivity. 

Best regards,
Marcel

**Comment by s-aibar** [comment]
Hello,
I am waiting for the response about the update of the R/Bioconductor version on the servers. If I don't get a reply within a couple more days, I will just do it on a local setup. Again, sorry for the delay...

Regards, 
Sara

**Comment by LiNk-NY** [comment]
Hi Sara, @s-aibar

The R version on the servers has been updated for quite some time. 
It is currently at R version `3.5` and Bioconductor version `3.7`.

Please push a version bump for the package to build. 
Thank you. 

Regards,
Marcel

**Comment by LiNk-NY** [comment]
Hi Sara, @s-aibar 

Please see here for the most current R version on the builders. 
http://www.bioconductor.org/checkResults/devel/bioc-LATEST/

Regards,
Marcel

**Comment by LiNk-NY** [comment]
Hi Sara, @s-aibar 

It seems to me like there was a misunderstanding here. You don't need to wait on your _institution_
to update their version of R, we will test the package with our version of R.

You can leave your depends field as `Depends: R (>= 3.2.0)` so that the package is compatible with
your _institution's_ R version. 

Just be aware that your package will be tested with R version `3.5` on the Bioconductor Build System (BBS) and that you should update your `Depends` field when your _institution's_ server has
been updated. 

Please apply the changes to the package and bump the version to trigger another build. 

Best regards,
Marcel

**Comment by LiNk-NY** [comment]
Hi Sara, @s-aibar

Please respond promptly or I will have to close the issue due to inactivity
(no commits since Oct. 24 aertslab/RcisTarget@92a791d8b09ca7e7ae6f8134c86e41db2b58018c).

Regards,
Marcel

**Comment by s-aibar** [comment]
Hi,
Sorry, Indeed it seems there was a misunderstanding with the need to update the version on our institution servers (In any case, I didn't expect it would take so long, otherwise I would have not waited, sorry about that...).
I will submit the pushes on the week of Jan 8.

**Comment by LiNk-NY** [comment]
Hi Sara, @s-aibar 

Just checking in to see how things are going. 

Best regards,
Marcel

**Comment by s-aibar** [comment]
Hello,

I have finally pushed the changes.
They include most of the suggestions from your last review (i.e. pre-allocate space for objects that grow, replace the fakeDatabase function for a saved objects in doc/examples, R version 3.4. The warning in the build is because I have set as requirement R 3.4 instead of 3.5...).

Regards,
Sara

**Comment by LiNk-NY** [comment]
Hi Sara, @s-aibar 

Thanks for making those changes. 
I had another look at your package and it looks almost ready to go. 

Please see this note in `BiocCheck`: 

```
Found <<- in R/priv_calcErn.R (line 71, column 47)
 * NOTE: Avoid '<<-' if possible (found in 1 files)
```

Can this use of `<<-` be avoided?

Thanks,
Marcel

**Comment by s-aibar** [comment]
Yes, it was possible to change it. I have replaced it by a for loop.

Thanks for having another look at the package :-) 
Sara

**Comment by LiNk-NY** [comment]
Hi Sara, @s-aibar 
Thanks for making that change. Your package has been approved.

Thank you for submitting to Bioconductor. 

Regards,
Marcel 

PS. Please remove the empty `Readme.md` from the data package.

**Comment by mtmorgan** [comment]
The master branch of your GitHub repository has been added to Bioconductor's git repository.

To use the git.bioconductor.org repository, we need an 'ssh' key to associate with your github user name. If your GitHub account already has ssh public keys (https://github.com/s-aibar.keys is not empty), then no further steps are required. Otherwise, do the following:

1. [Add an SSH key to your github account][]
1. [Submit your SSH key to Bioconductor][]

See further instructions at

https://bioconductor.org/developers/how-to/git/

for working with this repository. See especially

https://bioconductor.org/developers/how-to/git/new-package-workflow/
https://bioconductor.org/developers/how-to/git/sync-existing-repositories/

to keep your GitHub and Bioconductor repositories in sync.

Your package will be included in the next nigthly 'devel' build (check-out from git at about 6 pm Eastern; build completion around 2pm Eastern the next day) at

https://bioconductor.org/checkResults/

(Builds sometimes fail, so ensure that the date stamps on the main landing page are consistent with the addition of your package). Once the package builds successfully, you package will be available for download in the 'Devel' version of Bioconductor using `biocLite(\"RcisTarget.hg19.motifDBs.cisbpOnly.500bp\")`. The package 'landing page' will be created at

https://bioconductor.org/packages/RcisTarget.hg19.motifDBs.cisbpOnly.500bp

If you have any questions, please contact the bioc-devel mailing list (https://stat.ethz.ch/mailman/listinfo/bioc-devel); this issue will not be monitored further.

[Add an SSH key to your github account]: https://help.github.com/articles/connecting-to-github-with-ssh/
[Submit your SSH key to Bioconductor]: https://docs.google.com/forms/d/e/1FAIpQLSdlTbNjsQJDp0BA480vo4tNufs0ziNyNmexegNZgNieIovbAA/viewform

**Comment by mtmorgan** [comment]
The master branch of your GitHub repository has been added to Bioconductor's git repository.

To use the git.bioconductor.org repository, we need an 'ssh' key to associate with your github user name. If your GitHub account already has ssh public keys (https://github.com/s-aibar.keys is not empty), then no further steps are required. Otherwise, do the following:

1. [Add an SSH key to your github account][]
1. [Submit your SSH key to Bioconductor][]

See further instructions at

https://bioconductor.org/developers/how-to/git/

for working with this repository. See especially

https://bioconductor.org/developers/how-to/git/new-package-workflow/
https://bioconductor.org/developers/how-to/git/sync-existing-repositories/

to keep your GitHub and Bioconductor repositories in sync.

Your package will be included in the next nigthly 'devel' build (check-out from git at about 6 pm Eastern; build completion around 2pm Eastern the next day) at

https://bioconductor.org/checkResults/

(Builds sometimes fail, so ensure that the date stamps on the main landing page are consistent with the addition of your package). Once the package builds successfully, you package will be available for download in the 'Devel' version of Bioconductor using `biocLite(\"RcisTarget\")`. The package 'landing page' will be created at

https://bioconductor.org/packages/RcisTarget

If you have any questions, please contact the bioc-devel mailing list (https://stat.ethz.ch/mailman/listinfo/bioc-devel); this issue will not be monitored further.

[Add an SSH key to your github account]: https://help.github.com/articles/connecting-to-github-with-ssh/
[Submit your SSH key to Bioconductor]: https://docs.google.com/forms/d/e/1FAIpQLSdlTbNjsQJDp0BA480vo4tNufs0ziNyNmexegNZgNieIovbAA/viewform
